src/catchup/CatchupRange.cpp

```diff
@@ -37,4 +39,45 @@ operator!=(CatchupRange const& x, CatchupRange const& y)
 {
     return !(x == y);
 }
+
+CatchupRange
+makeCatchupRange(ResolvedCatchupConfiguration const& configuration,
+                 HistoryManager const& historyManager)
+{
+    auto toCheckpoint =
+        historyManager.nextCheckpointLedger(configuration.toLedger() + 1) - 1;
+
+    auto firstNotKnownCheckpoint = [&]() {
+        auto lcl = configuration.lastClosedLedger();
+        auto nextCheckpoint = historyManager.nextCheckpointLedger(lcl + 1) - 1;
+        if (lcl == nextCheckpoint)
+        {
+            return nextCheckpoint + historyManager.getCheckpointFrequency();
+        }
+        return nextCheckpoint;
+    }();
+
+    auto catchupFromLedger =
+        configuration.toLedger() > configuration.count()
+            ? configuration.toLedger() - configuration.count()
+            : 0;
+    auto catchupFromCheckpoint =
+        historyManager.nextCheckpointLedger(catchupFromLedger) - 1;
+
+    // ensure that we have at least configuration.count() ledgers if
+    // possible
+    if (catchupFromLedger < catchupFromCheckpoint &&
+        catchupFromCheckpoint > historyManager.getCheckpointFrequency())
+    {
+        catchupFromCheckpoint -= historyManager.getCheckpointFrequency();
+    }
+
+    auto fromCheckpoint =
+        std::max(catchupFromCheckpoint, firstNotKnownCheckpoint);
+
+    auto checkpointRange = CheckpointRange{fromCheckpoint, toCheckpoint};
+    auto applyBuckets = fromCheckpoint > firstNotKnownCheckpoint;
```

Comment:
I am not sure I follow how `applyBuckets` is computed, shouldn't it be related not to the firstNotKnownCheckpoint but instead just the LCL, ie if we can't replay tx directly, we have to jump to a place in history such that we have at least `configuration.count()` ledgers worth of transactions being replayed?

Reply:
Well, that is complicated.

firstNotKnownCheckpoint is related to LCL.

If we have for example LCL = 1022, then firstNotKnownCheckpoint is 1023 (as this is the first checkpoint that is not complety known).

Then fromCheckpoint value will be at least 1023 (thanks to the std::max) function. If it is equal to 1023 then we don't need to apply buckets - we will get transactions for checkpoints from 1023 to whatever toCheckpoint is. Transactions for checkpoint 1023 are: all from 958 to 1023, so we will just apply all of them (and skip all but last as implemented in ApplyLedgerChainWork).

If however fromCheckpoint is bigger, like 1087, then we need to apply buckets at 1087, as replaying ledgers for that checkpoint (1024 to 1087) would not work, with 1023 missing in our database.